### PR TITLE
move metrics to router: avoids headers not being forwarded and caching -...

### PR DESCRIPTION
**Pull request by zaach**
... issue #2040

### CODE REVIEW

File: bin/router [hunk 1]
```diff
@@ -146,6 +147,10 @@ wsapi.routeSetup(app, {
 
 //catch-all
 app.use(function(req, res, next) {
+
+  // log metrics
+  if (req.url === '/sign_in') metrics.userEntry(req);
```
Comment (lloyd) [inline bin/router:152]: nice, note to readers, this change will fix our under-reported metrics.

@zaach - should we log even if get data is present?

File: lib/static/views.js [hunk 2]
```diff
@@ -91,7 +91,6 @@ exports.setup = function(app) {
   // this should probably be an internal redirect
   // as soon as relative paths are figured out.
   app.get('/sign_in', function(req, res, next ) {
-    metrics.userEntry(req);
```
Comment (lloyd) [inline lib/static/views.js:94]: can we remove the require('metrics') line at the top of this file now?

File: lib/metrics.js [hunk 3]
```diff
@@ -56,6 +56,9 @@ function setupLogger() {
     mkdir_p(log_path);
 
   var filename = path.join(log_path, configuration.get('process_type') + "-metrics.json");
+  if (process.env.METRICS_LOG_FILE) {
+    filename = process.env.METRICS_LOG_FILE;
+  }
```
Comment (lloyd) [inline lib/metrics.js:61]: great approach to make this thing testable.

File: tests/metrics-header-test.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    },
+    "metrics log exists": {
+      topic: function (err, r) {
+        if (existsSync(process.env.METRICS_LOG_FILE)) {
+          this.callback();
+        } else {
+          fs.watchFile(process.env.METRICS_LOG_FILE, null, this.callback);
+        }
+      },
+      "metric fields are logged properly": function (event, filename) {
```
Comment (lloyd) [inline tests/metrics-header-test.js:72]: this is a beautiful test to ensure we don't regress.

**Comment by lloyd** [comment]
two nits, shouldn't block merge, this is great.  r+

note @vthunder - our metrics are fixed with this change which should go into production on august 1st.  Also, our metrics would have been under-reported.  this PR fixes that.  

Finally, note that we've change the name of the metrics file here, we should make sure to get that in the deployment ticket when we cut the train.

File: lib/static/views.js (hunk 2 above)
Comment (zaach) [inline lib/static/views.js:94]: Yep.

**Pull request merged**